operator/cmd/mesh/uninstall.go

```diff
@@ -103,50 +102,42 @@ func uninstall(cmd *cobra.Command, rootArgs *rootArgs, uiArgs *uninstallArgs, lo
 		return err
 	}
 
-	if err := uninstallWrapper(cmd, rootArgs, uiArgs, restConfig, clt, l); err != nil {
-		return fmt.Errorf("failed to uninstall control plane: %v", err)
-	}
-	return nil
-}
-
-// uninstallWrapper is helper function to uninstalls control plane by either
-// pruning by target revision or deleting specified manifests.
-func uninstallWrapper(cmd *cobra.Command, rootArgs *rootArgs, uiArgs *uninstallArgs,
-	restConfig *rest.Config, clt client.Client, l clog.Logger) error {
 	cache.FlushObjectCaches()
 	opts := &helmreconciler.Options{DryRun: rootArgs.dryRun, Log: l, ProgressLog: progress.NewLog()}
 	h, err := helmreconciler.NewHelmReconciler(clt, restConfig, nil, opts)
 	if err != nil {
 		return fmt.Errorf("failed to create reconciler: %v", err)
 	}
 
-	if uiArgs.revision != "" {
+	// If only revision flag is set, we would prune resources by the revision label.
+	// Otherwise we would merge the revision flag and the filename flag and delete resources by generated manifests.
+	if uiArgs.revision != "" && uiArgs.filename == "" {
 		objectsList, tp, err := h.GetPrunedResourcesList(uiArgs.revision)
 		if err != nil {
 			return err
 		}
-		if err := preCheckWarnings(cmd, rootArgs, uiArgs, uiArgs.revision, tp); err != nil {
+		if err := preCheckWarnings(cmd, rootArgs, uiArgs, uiArgs.revision, tp, l); err != nil {
 			return fmt.Errorf("failed to do preuninstall check: %v", err)
 		}
 
 		if err := h.PruneControlPlaneByRevision(uiArgs.revision, objectsList); err != nil {
 			return fmt.Errorf("failed to prune control plane resources by revision: %v", err)
 		}
-	} else if uiArgs.filename != "" {
-		manifestMap, iops, err := manifest.GenManifests([]string{uiArgs.filename}, applyFlagAliases(uiArgs.set, uiArgs.manifestsPath, uiArgs.revision), uiArgs.force, restConfig, l)
-		if err != nil {
-			return err
-		}
-		if iops == nil {
-			return fmt.Errorf("istioOperatorSpec is nil")
-		}
-		if err := preCheckWarnings(cmd, rootArgs, uiArgs, iops.Revision, nil); err != nil {
-			return fmt.Errorf("failed to do preuninstall check: %v", err)
-		}
-		cpManifests := manifestMap[name.PilotComponentName]
-		if err := h.DeleteControlPlaneByManifests(strings.Join(cpManifests, "---"), iops.Revision); err != nil {
-			return fmt.Errorf("failed to delete control plane by manifests: %v", err)
-		}
+	}
+	uiArgs.set = append(uiArgs.set, fmt.Sprintf("revision=%s", uiArgs.revision))
```

Comment:
this is unnecessary. if you look at manifest-apply.go runApplyCmd(), revision is parsed there as an alias and added to the list of set flags - you can just copy that pattern.